tests/e2e/test_read.py

```diff
@@ -39,29 +44,100 @@ def container_client(account_url):
+    blob_client = container_client.get_blob_client(blob=blob_name)
+    blob_client.upload_blob(data)
+    url = f"{account_url}/{container_client.container_name}/{blob_name}"
+
+    return Blob(data=data, url=url)
+
+
+@pytest.fixture(scope="module")
+def small_with_newlines_uploaded_blob(account_url, container_client):
+    blob_name = random_resource_name()
+    data = sample_data(20, True)
+    blob_client = container_client.get_blob_client(blob=blob_name)
+    blob_client.upload_blob(data)
+    url = f"{account_url}/{container_client.container_name}/{blob_name}"
+
+    return Blob(data=data, url=url)
 
 
 @pytest.fixture
-def blob_io(blob_url):
-    yield BlobIO(blob_url=blob_url, mode="rb")
+def uploaded_blob(request):
+    if request.param == "small_blob":
+        return request.getfixturevalue("small_uploaded_blob")
+    elif request.param == "large_blob":
+        return request.getfixturevalue("large_uploaded_blob")
+    elif request.param == "small_with_newlines_blob":
+        return request.getfixturevalue("small_with_newlines_uploaded_blob")
 
 
 def random_resource_name(name_length=8):
-    return "".join(random.choices(string.ascii_lowercase + string.digits, k=name_length))
+    return "".join(
+        random.choices(string.ascii_lowercase + string.digits, k=name_length)
+    )
+
+
+def sample_data(data_length=20, newline=False):
+    if newline:
+        return "".join(
+            random.choices(string.ascii_letters + string.digits + "\n", k=data_length)
+        ).encode()
+    return os.urandom(data_length)
 
 
 class TestRead:
-    def test_reads_all_data(self, blob_io, sample_data):
-        with blob_io as f:
-            assert f.read() == sample_data
+    @pytest.mark.parametrize(
+        "uploaded_blob",
+        [
+            "small_blob",
+            "large_blob",
+        ],
+        indirect=True,
+    )
+    def test_reads_all_data(self, uploaded_blob):
+        blob = uploaded_blob
+        with BlobIO(blob.url, "rb") as f:
+            assert f.read() == blob.data
+
+    @pytest.mark.parametrize("n", [1, 2, 4, 5, 9])
+    @pytest.mark.parametrize("uploaded_blob", ["small_blob"], indirect=True)
+    def test_read_n_bytes(self, uploaded_blob, n):
+        blob = uploaded_blob
+        with BlobIO(blob.url, "rb") as f:
+            for i in range(0, 10, n):
+                assert f.read(n) == blob.data[i : i + n]
+
+    @pytest.mark.parametrize("n", [1, 2, 4, 5, 9])
+    @pytest.mark.parametrize("uploaded_blob", ["small_blob"], indirect=True)
+    def test_random_seeks_and_reads(self, uploaded_blob, n):
+        blob = uploaded_blob
+        with BlobIO(blob.url, "rb") as f:
+            f.seek(n)
+            assert f.read() == blob.data[n:]
+
+    @pytest.mark.parametrize(
+        "uploaded_blob", ["small_with_newlines_blob"], indirect=True
+    )
+    def test_read_using_iter(self, uploaded_blob):
+        blob = uploaded_blob
+        with BlobIO(blob.url, "rb") as f:
+            data = b""
```

Comment:
For the iter test, it would be interesting to just collect each individual line in a list and we append each iteration to this list and from there we can just compare against splitting of the original data on newlines. The reason I'm suggesting it is that it is a stronger assertion that we are only returning a line for iteration of the for loop and we are currently only looking at the entire data content.